pkg/clustermesh/common/namespacewatcher/watcher.go

```diff
@@ -0,0 +1,507 @@
+}) bool {
+	// First check if service has the global annotation
+	if !annotation.GetAnnotationIncludeExternal(obj) {
+		return false
+	}
+
+	// If namespace filtering is active, also check if service is in a global namespace
+	if nw.IsFilteringActive() {
+		if !nw.IsGlobalNamespace(obj.GetNamespace()) {
+			// Service is marked as global but not in a global namespace
+			return false
+		}
+	}
+
+	return true
+}
+
+func (nw *namespaceWatcher) IsSharedGlobalService(obj interface {
+	GetAnnotations() map[string]string
+	GetNamespace() string
+}) bool {
+	// Service must be global first
+	if !nw.IsGlobalService(obj) {
+		return false
+	}
+
+	// Then check if it's also shared
+	return annotation.GetAnnotationShared(obj)
+}
+
+func (nw *namespaceWatcher) processNamespaceChange(processors []NamespaceProcessor, namespace string, isGlobal bool) {
+	if nw.syncProcessing {
+		// For testing: process synchronously
+		nw.processNamespaceChangeSync(processors, namespace, isGlobal)
+	} else {
+		// Normal operation: process asynchronously
+		for _, processor := range processors {
+			go processor.ProcessNamespaceChange(namespace, isGlobal)
+		}
+	}
+}
+
+// processNamespaceChangeSync is a synchronous version for testing
+func (nw *namespaceWatcher) processNamespaceChangeSync(processors []NamespaceProcessor, namespace string, isGlobal bool) {
+	// Process the namespace change synchronously for testing
+	for _, processor := range processors {
+		processor.ProcessNamespaceChange(namespace, isGlobal)
+	}
+}
+
+func (nw *namespaceWatcher) updateNamespace(ns *slim_corev1.Namespace) {
+	// Check if namespace has the global annotation
+	annotationValue, hasAnnotation := annotation.Get(ns, annotation.GlobalNamespace)
+
+	nw.mu.Lock()
+
+	// Capture the previous state by querying the resource store
+	wasAnnotated := nw.isNamespaceAnnotated(ns.Name)
+	wasGlobal := nw.isNamespaceGlobalByAnnotation(ns.Name)
+	wasFilteringActive := nw.isFilteringActive()
+
+	var processors = make([]NamespaceProcessor, len(nw.processors))
+	copy(processors, nw.processors)
+
+	nw.mu.Unlock()
+
+	// Calculate current global status after the namespace update
+	// We need to check the current filtering state and annotation status
+	isFilteringActiveNow := hasAnnotation || nw.hasOtherAnnotatedNamespaces(ns.Name)
+	isGlobal := nw.calculateGlobalStatus(ns.Name, hasAnnotation, annotationValue, isFilteringActiveNow)
+
+	// Check if this specific namespace's status changed
+	wasGlobalWithOldLogic := wasAnnotated && wasGlobal || (!wasAnnotated && nw.config.DefaultGlobalNamespace)
+	needsProcessing := (wasGlobalWithOldLogic != isGlobal)
+
+	// Check if we transitioned from filtering active to inactive (backfill case)
+	transitionedToInactive := wasFilteringActive && !isFilteringActiveNow
+
+	// Check if we transitioned from filtering inactive to active (cleanup case)
+	transitionedToActive := !wasFilteringActive && isFilteringActiveNow
+
+	if needsProcessing {
+		if nw.logger != nil {
```

Comment:
Is this just for testing? Wouldn't it be simple to just expect to use a test logger instead of having this check everywhere?

Reply:
Done